src/PaginatedData.elm

```diff
@@ -1,353 +1,416 @@
+
+
+### Accessors
+
+@docs get, getAll, getItemsByPage, getTotalCount, getPage, getLocal
+
+
+### Fetch helpers
 
-{-| A `PaginatedData` represents a dict of values, that are paginated on the
-server.
+@docs fetchAllPages, fetchNextPage
 
-@docs ContainerDict, PaginatedData, emptyContainer, emptyPaginatedData, fetchAll, fetchPaginated, get, getAll, getItemsByPager, getPager, getTotalCount, insertDirectlyFromClient, insertMultiple, remove, setPageAsLoading, setTotalCount, update, viewPager
+
+### Updaters
+
+@docs insertLocal, handleFetchedPage, remove, setPageAsLoading, setTotalCount, update
+
+
+### View
+
+@docs viewPager
 
 -}
 
-import EveryDict exposing (EveryDict)
+import Dict exposing (Dict)
 import EveryDictList exposing (EveryDictList)
 import Html exposing (Html, a, li, text, ul)
 import Html.Attributes exposing (action, class, classList)
 import Html.Events exposing (onClick)
-import RemoteData exposing (WebData)
+import Http
+import RemoteData exposing (RemoteData(..))
 
 
-{-| A container Dict can act as a local cache.
+{-| Represents the status of paged data, where we may have some or all of the
+pages, and requests for one or more pages may be in-flight or have had errors.
+You may also have some "local" data which is not on any page.
 
-@todo: Add docs.
+The `value` is the type of the value that you are fetching in pages.
 
--}
-type alias ContainerDict identifier key value =
-    EveryDict identifier (WebData (PaginatedData key value))
+The `key` is a type for looking up values. For instance, it might be an ID for
+the values. It is used as the key for `EveryDictList`, so it should be unique,
+and it must have work with `toString` in a way that produces unique results.
 
+The `err` is a type for tracking errors fetching pages. For the usual case
+of Http requests, this will be `Http.Error`. The `PaginatedWebData` type
+is a convenient alias for that common case.
 
-{-| We need to know how much pages we have so we could lazy load.
-The pager holds the a tuple with the first and last Item key for that page, so
-it's easier to insert new items in the correct place.
 -}
-type alias PaginatedData key value =
-    { data : EveryDictList key value
-    , pager : EveryDict Int (WebData ( key, key ))
+type PaginatedData err key value
+    = PaginatedData
+        -- A record representing an attempt to fetch each page. In
+        -- `handleFetchedPage`, we look at the `totalCount` and try to infer the
+        -- page size, so we will actually have a `NotAsked` entry here for
+        -- every page we expect to exist.
+        { pager : Pager err key value
+
+        -- Represents the number of items which we would have once we receive
+        -- all the pages, if we know that. (For instance, the backend might
+        -- tell us how many items there in total are while returning just one
+        -- page). Note that this is the item count, not the page count.
+        , totalCount : Maybe Int
```

Comment:
since total count is related to the pager, I wonder if it should not be moved to be a part of it. Also, maybe we should rename `pager` to `backend`, given the `local` property?

Someting like:

``` elm
{ backend : (Pager err key value, Maybe Int)`
, local : EveryDictList key value
}
```